src/dotnet/UI.Blazor/Services/HistoryUI/HistoryUI.cs

```diff
@@ -0,0 +1,228 @@
+
+    public event EventHandler<AfterLocationChangedHandledEventsArgs>? AfterLocationChangedHandled;
+
+    public HistoryUI(
+        BrowserInfo browserInfo,
+        IJSRuntime js,
+        NavigationManager nav,
+        ILogger<HistoryUI> log)
+    {
+        BrowserInfo = browserInfo;
+        JS = js;
+        Log = log;
+        Nav = nav;
+
+        InitialLocation = Nav.Uri;
+        IsInitialLocation = true;
+        _historyIndex = 0;
+        _history.Add(new HistoryItem(InitialLocation));
+        Log.LogDebug("Initial location: '{Location}'", Nav.Uri);
+
+        // HistoryUI is initialized upon BlazorCircuitContext is created.
+        // At this moment there is no yet subscribers to Nav.LocationChanged.
+        // So HistoryUI will be notified the first on location changed, even before the Router.
+        Nav.LocationChanged += OnLocationChanged;
+
+        _whenLocationChangedHandled = Task.CompletedTask;
+        _whenJsRefInitialized = InitializeJsRef();
+        _state = new State { Index = 0 };
+        WhenInitialized = InitializeState(_state);
+    }
+
+    public void Dispose()
+        => _blazorRef?.Dispose();
+
+    [JSInvokable]
+    public Task OnPopState(string state)
+        => Task.CompletedTask;
+
+    public async Task RouterOnNavigateAsync(NavigationContext arg)
+        // Postpone Router navigation till location changed async completed.
+        // We can get rid of this event handler
+        // after OnLocationChangedAsync is replaced with synchronous implementation in .NET 7.
+        => await _whenLocationChangedHandled;
+
+    public Task GoBack()
+        => JS.InvokeVoidAsync("eval", "history.back()").AsTask();
+
+    public void NavigateTo(Action? onForwardAction, Action? onBackAction)
+    {
+        if (_pendingHistoryItem != null)
+            throw StandardError.Constraint("There is still pending history item");
+        var historyItem = new HistoryItem(Nav.Uri) {
+            OnForwardAction = onForwardAction,
+            OnBackAction = onBackAction
+        };
+        _pendingHistoryItem = new PendingHistoryItem(historyItem, _historyIndex);
+        Nav.NavigateTo(Nav.Uri);
+    }
+
+    private async Task<State?> GetStateAsync()
+    {
+        await _whenJsRefInitialized.ConfigureAwait(false);
+        var result = await _jsRef!.InvokeAsync<State?>("getState").ConfigureAwait(false);
+        return result;
+    }
+
+    private async Task InitializeJsRef()
+    {
+        _blazorRef = DotNetObjectReference.Create<IHistoryUIBackend>(this);
+        _jsRef = await JS.InvokeAsync<IJSObjectReference>(
+            $"{BlazorUICoreModule.ImportName}.HistoryUI.create",
+            _blazorRef).ConfigureAwait(false);
+    }
+
+    private async Task InitializeState(State state)
+    {
+        await _whenJsRefInitialized;
+        await SetStateAsync(state);
+    }
+
+    private void OnLocationChanged(object? sender, LocationChangedEventArgs e)
+    {
+        IsInitialLocation = false;
```

Comment:
even assuming that we need this flag, can't it be calculated like InitialLocation == currentLocation && historyIndex== 0/1? not suggesting, just asking

Reply:
No, it indicates that we are excatly on the first location and did not move anywhere yet.
I think now I can rework it with using _history, but this will be less clear.

Reply:
ok then, let's leave explicit set